rust/rope/src/tree.rs

```diff
@@ -763,25 +713,35 @@ impl<'a, N: NodeInfo> Cursor<'a, N> {
         CursorIter { cursor: self, _metric: PhantomData }
     }
 
+    /// Tries to find the last boundary in the leaf the cursor is currently in.
+    ///
+    /// If the last boundary is at the end of the leaf, it is only counted if
+    /// it is less than `orig_pos`.
+    #[inline]
+    fn last_inside_leaf<M: Metric<N>>(&mut self, orig_pos: usize) -> Option<usize> {
+        let l = self.leaf.expect("inconsistent, shouldn't get here");
+        let len = l.len();
+        if self.offset_of_leaf + len < orig_pos && M::is_boundary(l, len) {
+            let _ = self.next_leaf();
```

Comment:
I'm slightly confused, here. `self.next_leaf()` modifies `self.position`; we want to be returning `self.offset_of_leaf + len`, I think; I can't see that this is equivalent?

Reply:
This handles specifically the case where the boundary is at the end of the current leaf. To satisfy the invariants, we advance to the start of the next leaf.